### PR TITLE
Adding convection diffusion example in C++

### PR DESCRIPTION
**Pull request by SavannaSpayd**
PR for Issue #25

### CODE REVIEW

**Comment by valeriabarra** [comment]
Hi @SavannaSpayd , thank you so much for your work! 

We generally do not want to submit to our PRs the files automatically generated by IDEs (such as VSCode). This should have been caught by the `.gitignore`, but for some reason it didn't. 

Please remove the `.DS_Store` file from the PR. Thank you!

File: examples/examples_C++/convection_diffusion.cpp
Comment (valeriabarra) [inline examples/examples_C++/convection_diffusion.cpp:1]: Hi @SavannaSpayd  thank you so much for considering contributing to this community project. Any contributions are welcome! However, you seem to have submitted code that does not use any functionality of the MOLE library. That is not the intent of adding an example in the examples directory. We want examples that show the functionality of the library, so you need to use the MOLE built-in functions in your code.

You could still use this code for comparison purposes on your local machine, to verify the correctness of your MOLE example. But please do not submit this code to the PR. 

Also you seem to have checked out your branch from an old version of `master`. The directory tree has now changed on `master`, so now the new path for C++ examples is `examples/cpp`.

There are two ways to update your branch with the current master branch.

1. The [git merge](https://git-scm.com/docs/git-merge) way: Update the master branch by doing git checkout master and then git pull. Once master is up to date, you can merge the master branch into yours by running git checkout your_branch (i.e., go back to your own branch) and then git merge master. This way will merge all commits there are currently on master into your own feature branch. This way causes the commit history of branch to be mingled with your own local feature branch commit history. For this reason, I tend to avoid this way.

2. The [git rebase](https://git-scm.com/docs/git-rebase) way. Update the master branch by doing git checkout master and then git pull. Once master is up to date, you can rebase your branch on top of the master branch by running git checkout your_branch (i.e., go back to your own branch) and then git rebase master. This way your feature branch commit history will always be on top of the master branch commit history. It is a lot cleaner and you can easily identify your own commits via git log and change commit history if needed.


I hope this helps!

**Comment by valeriabarra** [comment]
Hello @SavannaSpayd , in addition to my comments in the PR Review that I hope you are working on, please note that we want to make sure that your C++ exactly reproduces (or at least up to a good amount of significant digits) the corresponding MATLAB example. Did you execute the MATLAB example, printed its numerical values to file and compared with the numbers you get from your C++ example? Please do that as verification sanity-check of your code and post a screenshot of the comparison here in the PR conversation. Thank you!

**Comment by SavannaSpayd** [comment]
@valeriabarra Thank you so much for the clarification! 

I've gone and removed that pesky .DS_Store file, as well as updated the convection_diffusion.cpp to properly utilize the mole library. 

In order to fully compare the matlab and c++ outputs to ensure they are the same, I have added an optional logging capability to the example c++ program that can output all of the data to a text file. This file can be easily swept to plot the concentration of CO2 over time, exactly as the matlab example as done. 

I've thoroughly tested the new c++ example, and have provided the output comparison of both the convection_diffusion.cpp and convection_diffusion.m below:


https://github.com/user-attachments/assets/897ce4ca-19be-4f57-bb13-203329cd0a40

https://github.com/user-attachments/assets/17a23c0d-716e-46d9-8e2c-5649942a256c

As you can see, the concentration over time is exactly what the matlab example has produced. 

Please let me know if you see any other issues with this PR, or would like any other additions. Thank you!

**Comment by valeriabarra** [comment]
Hi @SavannaSpayd, thank you so much for your work, addressing the comments and posting the videos! The videos look very good. To be extra sure, it would be great if you could do a sanity check and check the numerical values, just at the last time step (no need to compare all time steps). Can you please share here in the PR conversation what those numbers look like the for the Matlab and C++ codes?

Thank you

**Comment by SavannaSpayd** [comment]
@valeriabarra, sure thing. I locally modified the MATLAB script to output its last frame to a text file, so that I can compare the two programs' outputs. Here were the two outputs:

[frames.txt](https://github.com/user-attachments/files/18138812/frames.txt)
[matlab_last_frame.txt](https://github.com/user-attachments/files/18138801/matlab_last_frame.txt)

To compare these outputs, I developed a Python script that reads both files and checks each numerical value in Frame #360 (the final frame). The comparison revealed a margin of error of less than 1e-7 between the C++ and MATLAB versions. I believe this minor discrepancy is attributable to the floating-point precision limitations between the two languages.

Let me know if there's anything else that should be looked at as well.

Thank you.

**Comment by valeriabarra** [comment]
Hi @SavannaSpayd , thank you so much for addressing my latest comment. You also need to modify the `examples/cpp/Makfile` so that your file is automatically compiled. If you do not know how to do it, see one of your class mates PRs', e.g., [this one](https://github.com/csrc-sdsu/mole/pull/35/files), which had similar relevant changes to the Makefile. Thank you.

**Comment by SavannaSpayd** [comment]
@valeriabarra Thanks so much for the heads up, makefile should be updated :)

**Review by valeriabarra — APPROVED** [review]

**Pull request merged**
